rclcpp/include/rclcpp/node_impl.hpp

```diff
@@ -235,14 +235,38 @@ Node::set_parameters_atomically(
 {
   std::lock_guard<std::mutex> lock(mutex_);
   std::map<std::string, rclcpp::parameter::ParameterVariant> tmp_map;
+  std::vector<rcl_interfaces::msg::Parameter> new_parameters;
+  std::vector<rcl_interfaces::msg::Parameter> changed_parameters;
+  std::vector<rcl_interfaces::msg::ParameterDescriptor> deleted_parameters;
+
   for (auto p : parameters) {
+    if (parameters_.find(p.get_name()) == parameters_.end()) {
+      new_parameters.push_back(p.to_parameter());
+    } else if (p.get_type() != rclcpp::parameter::ParameterType::PARAMETER_NOT_SET) {
+      changed_parameters.push_back(p.to_parameter());
+    } else {
+      rcl_interfaces::msg::ParameterDescriptor descriptor;
+      descriptor.name = p.get_name();
+      descriptor.parameter_type = p.get_type();
+      deleted_parameters.push_back(descriptor);
+    }
     tmp_map[p.get_name()] = p;
   }
+  // std::map::insert will not overwrite elements, so we'll keep the new
+  // ones and add only those that already exist in the Node's internal map
   tmp_map.insert(parameters_.begin(), parameters_.end());
   std::swap(tmp_map, parameters_);
+
   // TODO: handle parameter constraints
   rcl_interfaces::msg::SetParametersResult result;
   result.successful = true;
+
+  auto parameter_event = std::make_shared<rcl_interfaces::msg::ParameterEvent>();
```

Comment:
Instead of creating local variables above and then copying the data into the event instance the event instance could be created earlier and the members could be populated directly.